src/System.Slices/System/Buffers/OwnedMemory.cs

```diff
@@ -120,8 +122,8 @@ protected internal virtual DisposableReservation Reserve(ref ReadOnlyMemory<T> m
         #region Used by Memory<T>
         void IKnown.AddReference(long id)
         {
-            VerifyId(id);
             AddReference();
+            VerifyId(id);
```

Comment:
We could catch the exception, Decrement the reference count, and rethrow the exception?  Or unpack VerifyId if you don't want the rethrow.